### ISSUE TITLE
GetQuery with optional parameters

### ISSUE THREAD

**Issue by FerLuisxd** [uNetworking/uWebSockets.js#344]
Right now the getQuery function will ignore any parameter, woudln't it be nice to have the option to get the exact Query you are looking for?
Making it optional so it doesn't affect performance at all if no params is sent

**Comment by hst-m** [comment]
It's easy enough to do in JS, here is the 1 line function I use:

```javascript
const queryToObj = req=>req.getQuery().split('&').reduce((obj,x)=>{let [a,b]=x.split('=');obj[a]=b;return obj},{})

// example
uws.App().any('/*',(res,req)=>{
   const obj=queryToObj(req)
   console.log(obj.field1, obj.field2)
})
```

see once the handler is called for that url it's got the values for req already, it doesn't know if you are going to try to call getQuery with a parameter or not so it would be waste of cpu to parse the query string for every request, it is easy enough to do it in JS and if you need more custom parsing you can do it in JS

**Comment by alexhultman** [comment]
@FerLuisxd You want getQuery(name)? That makes sense, I haven't added much HTTP features.

> const queryToObj = req=>req.getQuery().split('&').reduce((obj,x)=>{let [a,b]=x.split('=');obj[a]=b;return obj},{})

This has to be one of the most inefficient lines of code I have ever seen in my life, honestly.

You scan linearly for &, allocate and fill an array that is garbage collected, then you reduce the array (a second pass), allocating yet another array of key, value for every entry, filling a hash map with those (key, value) that is also garbage collected, then you use that hash map to retrieve the value of a certain key.

This is algorithmical madness, you're invoking a shit ton of garbage collection and re-scan of the same data. res.getQuery(name) could be entirely zero-copy if done in the library.

**Comment by hst-m** [comment]
add a `getQuery(name)` and I will benchmark test it vs `queryToObj` and see what the difference is, I've got some large queries to use it on, I did not consider performance when writing that function more of a quick easy parse

**Comment by alexhultman** [comment]
You don't need to benchmark it, I can tell just by looking at it

**Comment by hst-m** [comment]
only 1 array was created from `getQuery().split('&')`, that array is reduced to a single object that gets returned by the function
> let [a,b]=x.split('=')

this does not create an array, it defines a and b and the code adds those to the single obj that was returned

**Comment by alexhultman** [comment]
Correction: all garbage collected languages are crap

**Comment by hst-m** [comment]
Just a regex lookup per field name is probably better performance, just need to re-scan the string every lookup. That's basically what the c lookup would do right, you think still difference in performance for that? I should switch my js to regex

**Comment by alexhultman** [comment]
Regex is pure crap as well. It's heavy and slow as shit.

**Comment by alexhultman** [comment]
Scanning for & and = is extremely cheap and is not going to take any measurable time whatsoever, esp. not for small strings like these (they cannot possibly be larger than 4kb). It literally cost you nothing at all, if you do it using a simple while loop in native code. You can't even measure the cost, in comparison.

The worst you can possibly do is bring a bunch of memory allocations to this. It's a really simple problem, you don't need any extra memory for it.

**Comment by hst-m** [comment]
I've got geo queries with shapes/coordinate paths that can be long but can just put those at the end of the query so it doesn't scan over those per lookup

**Comment by alexhultman** [comment]
I mean, this is already solved. The library already does a parse of Http headers this way.

You scan once, put the offsets on a stack, then when you retrieve a query you scan the stack of keys. That way you have no allocations, and the problem becomes pretty much free of cost.

It makes sense to do this by default, as soon as there is a query string because the user will want to read it anyways.

**Comment by alexhultman** [comment]
You can also do a slight modification to make it lazy:

getQuery(name) looks in the stack of already scanned keys, then if not found continues to scan the querystring until it finds it or end. But for every found query that is not the one we look for, add it to the stack of already scanned keys.

This way it becomes free of cost when you don't use getQuery(name), while still being fast enough for the cases when you do. I think this is the way to go.

**Comment by FerLuisxd** [comment]
> I mean, this is already solved. The library already does a parse of Http headers this way.
> 
> You scan once, put the offsets on a stack, then when you retrieve a query you scan the stack of keys. That way you have no allocations, and the problem becomes pretty much free of cost.
> 
> It makes sense to do this by default, as soon as there is a query string because the user will want to read it anyways.

This will need a change to uWebSockets.js and uWebSockets, right?  So it will need a change https://github.com/uNetworking/uWebSockets.js/blob/7da3a0a787ad2a6f59b0611a3d5052b88e99af92/src/HttpRequestWrapper.h#L99 and https://github.com/uNetworking/uWebSockets/blob/7d31bad8f2724a79b1346f0d5c4a30a8372ef221/src/HttpParser.h#L111

**Comment by hst-m** [comment]
my first example was doing it in the least amount of code (I have tendency to do that over performance, not ideal) here is how I would do it in JS for the best performance, I am switching to this in my apps for now:

```javascript
const queryToObj = req => {
   const obj = {}, string = req.getQuery()
   let index = 0, index2, index3 = 0
   do {
      index2 = string.indexOf('=', index3+1)
      if(index2 == -1) return obj 
      index3 = string.indexOf('&', index2+1)
      if(index3 == -1) index3 = string.length
      obj[string.slice(index, index2)] = string.slice(index2+1, index3)
      index = index3+1
   } while (index3 != string.length)
   return obj
}
```

```javascript
// example
uws.App().get('/', (res,req) => {
   const obj = queryToObj(req)
   console.log(obj.field)
})
```
measured at about twice as fast as the previous method with one of my average queries, only scans each letter once with `indexOf`, no extra arrays created etc just the single object. You could also add the "lazy" feature to stop when the requested field is found, don't slice the extra strings unless needed etc but I leave that out because I usually need all the parameters. Ideally the c library would have this in there because everyone needs it but can compare the performance for fun see if any faster

**Comment by uasan** [comment]
You need to decode the values, this method - `decodeURIComponent`

**Comment by hst-m** [comment]
@uasan if you can control/sanitize your query so that it doesnt use encoded values you can skip that, I just did a quick test and am showing a huge 10 times slower performance if you add those in, that is pretty big so would recommend skipping that if possible,  my endpoint specifically avoids encoded values but if you need it you can add it in, just know it will be much slower

**Comment by alexhultman** [comment]
Decoding an URI doesn't have to be slow if done properly. You just scan it once, building the decoded string as you go. This is something that should also be done for the router, so that you can have routes that have, say, Swedish letters or spaces.

Decoding is a scan for %, that's all.

**Comment by alexhultman** [comment]
I just wrote a super simple decode based on an assumption and it works really well. All you do is replace all %FF with their "binary" byte. That's it. It's really simple to do efficiently.

**Comment by hst-m** [comment]
@uasan if you know you have some values that need to be decoded you could do them individually outside the getQuery function, that would speed it up:
```javascript
const obj = getQuery(req)
decodeURIComponent(obj.field)
```

**Comment by uasan** [comment]
@hst-m yes, i understand you, I just pointed out that need not forget about decoding.
Your second option is acceptable - decode if only need

**Comment by hst-m** [comment]
the OP @FerLuisxd was basically asking for the "lazy" version, having an optional name argument that does not affect performance (or increases performance). I updated the previous example for that as described by @alexhultman and it works well, if your query field is early in the string it is like 15-20 times faster than parsing the whole string, and the state gets saved if you need to look up more fields. So you can use this to grab all the fields with `getQuery(req)` or get a specific field `getQuery(req, 'name')` and it will be faster and still save state for future lookups. So I recommend using this, and it does what OP requested, will this work for you @FerLuisxd? try this out:
```javascript
const getQuery = (req, name) => {
   if(!req.query) {
      req.query = { obj: {}, lastIndex: 0, string: req.getQuery() } 
   } else {
      if(req.query.obj[name]) return req.query.obj[name]
      if(req.query.lastIndex >= req.query.string.length) return name ? null : req.query.obj
   }
   let index2, index3
   do {
      index2 = req.query.string.indexOf('=', req.query.lastIndex)
      if(index2 == -1) index2 = req.query.string.length
      index3 = req.query.string.indexOf('&', index2+1)
      if(index3 == -1) index3 = req.query.string.length
      const key = req.query.string.slice(req.query.lastIndex, index2)
      req.query.obj[key] = req.query.string.slice(index2+1, index3)
      req.query.lastIndex = index3+1
      if(key == name) return req.query.obj[key]
   } while (index3 != req.query.string.length)
   return name ? null : req.query.obj
}
```

```javascript
// example

const value = getQuery(req, 'name')
const obj = getQuery(req)
```

**Comment by alexhultman** [comment]
str.includes('%') has the same algorithmic cost as doing the full decodeURIComponent. They both (if implemented correctly) does a linear scan, replacing any % signs it finds. This is linear in complexity for both. So by checking first with includes('%') you're doing potentially twice the work.

**Comment by alexhultman** [comment]
The point of this issue is to add req.getQuery(name) with proper URI decoding built-in. So any recommendation is going to be, well, use the native built-in feature when it lands. Until then, use whatever. But the built-in is going to be by far the most efficient (and simplest since it is built-in).

**Comment by hst-m** [comment]
you will be adding a '%' check for every character in every string even for those who have endpoints designed to not have encodings, this would only slow it down for that case? could have an 'auto-decode' option to turn it on or off?

**Comment by alexhultman-2fa** [comment]
Finding a % in a string smaller than 4kb is practically free, esp when you use SIMD instructions. We already parse HTTP this way. Most text protocols parse this way, that's why binary protocols like WebSocket is king
